api/src/main/java/org/projectnessie/tools/catalog/migration/api/CatalogMigrator.java

```diff
@@ -0,0 +1,277 @@
+      Catalog targetCatalog,
+      List<TableIdentifier> registeredTableIdentifiers,
+      List<TableIdentifier> failedToMigrateTableIdentifiers,
+      TableIdentifier tableIdentifier) {
+    try {
+      // register the table to the target catalog
+      TableOperations ops = ((BaseTable) sourceCatalog.loadTable(tableIdentifier)).operations();
+      targetCatalog.registerTable(tableIdentifier, ops.current().metadataFileLocation());
+
+      registeredTableIdentifiers.add(tableIdentifier);
+      LOG.info("Successfully migrated the table {}", tableIdentifier);
+      return true;
+    } catch (Exception ex) {
+      failedToMigrateTableIdentifiers.add(tableIdentifier);
+      LOG.warn("Unable to register the table {}", tableIdentifier, ex);
+      return false;
+    }
+  }
+
+  private static List<TableIdentifier> getMatchingTableIdentifiers(
+      Catalog sourceCatalog, String identifierRegex, PrintWriter printWriter) {
+    if (identifierRegex == null) {
+      printWriter.println(
+          String.format(
+              "%nUser has not specified the table identifiers."
+                  + " Selecting all the tables from all the namespaces from the source catalog."));
+    } else {
+      printWriter.println(
+          String.format(
+              "%nUser has not specified the table identifiers."
+                  + " Selecting all the tables from all the namespaces from the source catalog "
+                  + "which matches the regex pattern:"
+                  + identifierRegex));
+    }
+
+    printWriter.println("Collecting all the namespaces from source catalog...");
+    // fetch all the table identifiers from all the namespaces.
+    List<Namespace> namespaces =
+        (sourceCatalog instanceof SupportsNamespaces)
+            ? ((SupportsNamespaces) sourceCatalog).listNamespaces()
+            : ImmutableList.of(Namespace.empty());
+    if (identifierRegex == null) {
+      printWriter.println("Collecting all the tables from all the namespaces of source catalog...");
+    } else {
+      printWriter.println(
+          "Collecting all the tables from all the namespaces of source catalog"
+              + " which matches the regex pattern:"
+              + identifierRegex);
+    }
+
+    Predicate<TableIdentifier> matchedIdentifiersPredicate;
+    if (identifierRegex != null) {
+      Pattern pattern = Pattern.compile(identifierRegex);
+      matchedIdentifiersPredicate =
+          tableIdentifier -> pattern.matcher(tableIdentifier.toString()).matches();
+    } else {
+      matchedIdentifiersPredicate = tableIdentifier -> true;
+    }
+    return getMatchingTableIdentifiers(sourceCatalog, namespaces, matchedIdentifiersPredicate);
+  }
+
+  private static List<TableIdentifier> getMatchingTableIdentifiers(
+      Catalog sourceCatalog,
+      List<Namespace> namespaces,
+      Predicate<TableIdentifier> matchedIdentifiersPredicate) {
+    List<TableIdentifier> allIdentifiers = new ArrayList<>();
```

Comment:
It's a repeated mistake repeated over and over again in your PRs.
Unnecessary collection + copying + stream misuse.